timemachine/cpp/src/barostat.cu

```diff
@@ -0,0 +1,295 @@
+#include "barostat.hpp"
+#include "gpu_utils.cuh"
+#include "fixed_point.hpp"
+#include <algorithm>
+#include <stdio.h>
+
+#define AVOGADRO 6.0221367e23
+#define BOLTZ 0.008314462618
+
+namespace timemachine {
+
+MonteCarloBarostat::MonteCarloBarostat(
+    const int N,
+    const double pressure,  // Expected in Bar
+    const double temperature, // Kelvin
+    const std::vector<std::vector<int> > group_idxs,
+    const int interval,
+    const std::vector<BoundPotential *> bps,
+    const int seed) :
+    N_(N),
+    pressure_(pressure),
+    temperature_(temperature),
+    interval_(interval),
+    bps_(bps),
+    group_idxs_(group_idxs),
+    volume_scale_(0),
+    num_attempted_(0),
+    num_accepted_(0),
+    seed_(seed),
+    step_(0) {
+
+    // lets not have another facepalm moment again...
+    if(temperature < 100.0) {
+        std::cout << "warning temperature less than 100K" << std::endl;
+    }
+
+    if(pressure > 10.0) {
+        std::cout << "warning pressure more than 10bar" << std::endl;
+    }
+
+    mt_ = std::mt19937(seed_);
+    dist_ = std::uniform_real_distribution<double>(0.0, 1.0);
+
+    //gpuErrchk(cudaMalloc(&d_group_idxs_, group_idxs_.size()*sizeof(*d_group_idxs_)));
+
+    const int num_mols = group_idxs_.size();
+
+    gpuErrchk(cudaMalloc(&d_x_after_, N_*3*sizeof(*d_x_after_)));
+    gpuErrchk(cudaMalloc(&d_box_after_, 3*3*sizeof(*d_box_after_)));
+
+    gpuErrchk(cudaMalloc(&d_u_before_, N_*sizeof(*d_u_before_)));
+    gpuErrchk(cudaMalloc(&d_u_after_, N_*sizeof(*d_u_after_)));
+
+    gpuErrchk(cudaMalloc(&d_centroids_, num_mols*3*sizeof(*d_centroids_)));
+    gpuErrchk(cudaMalloc(&d_mol_idxs_, N_*sizeof(*d_mol_idxs_)));
+    gpuErrchk(cudaMalloc(&d_mol_offsets_, (num_mols+1)*sizeof(*d_mol_offsets_)));
+
+    int offset = 0;
+    int mol_offsets[num_mols+1];
+    int mol_idxs[N_];
```

Comment:
You probably don't want to use N_, but to recompute this based on the sum of group.size() in group_idxs...

Reply:
If N_ != sum([g.size() for g in group_idxs]) this will bork when indexing in the kernels, should I add an assertion that the two values are the same?